src/utility.cpp

```diff
@@ -0,0 +1,139 @@
+// Linux
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+// local headers
+#include "utility.h"
+
+// C++
+#include <iostream>
+#include <sstream>
+#include <string>
+
+bool existsFile(const std::string &path)
+{
```

Comment:
Well for one I personally dislike `access()` due to its racy nature. Secondly I dislike using it for existence tests since semantically `access()` checks for certain access rights and here we only want to know whether the file exists or not.

Given that, the error handling does not reflect that. The reasons probably is that the original code *did* use `access()` and I didn't want to change the behaviour.

Maybe having a function `canAccess()` and using `access()` in there would make more sense.

Reply:
I understand the source of dislike, but existence tests are racy by definition...

I think `access` is the syscall that does the least things in the kernel and still allows to check for existence - it doesn't touch the file descriptor table, doesn't retrieve all the data needed for `stat`, etc. Performance doesn't matter too much here, but that's the reason why I'd usually prefer it for this use-case.

We could also compromise to just remove the existence-checks and make sure that `parseProfile` handles missing profiles gracefully.

Reply:
I'd rather keep the existence checks as they are at the moment. They're rather deep in the program logic. The most robust approach would actually be to actually `open()` the files right away and keep the open file along the path. But that would also be a larger change.

Thinking about it once more I think an `access(..., F_OK)` will do nicely for the implementation.

After the basic refactoring is done we could also think about some sanity checks like rejecting files that have bad file ownership or permissions. This is something that happens at times that admins break their config files this way ...

Reply:
One further aspect has come up here. `access()` succeeds for both regular files and directories alike. Symlinks are dereferenced unconditionally. So in theory, if something is badly messed up, chkstat would try to open a directory instead of a file and fail. This is the original behaviour, however. Therefore I'll change this to `access(..,. R_OK)` for the time being. Even though I dislike it.

Reply:
Directories are interesting... `open` works on them, too. The only ways to check for them I can think of are to try to `read` from the resulting file descriptor and checking for `EBADFD`. Or `open(..., O_DIRECTORY)` and then checking for `ENOTDIR`.